### ISSUE TITLE
Syntax error: different prefix: "/" and "../media/img" for `sprite-file' 1.0.0.alpha19

### ISSUE THREAD

**Issue by jsonUK** [Compass/compass#1617]
Hi,

I currently get the error:

```
Syntax error: different prefix: "/" and "../media/img" for `sprite-file'
        on line 107 of ../src/sass/_mixins.scss, in `use-sprite'
        from line 43 of ../src/sass/core/_icons.scss
        from line 68 of ../src/sass/style.scss
  Use --trace for backtrace.
```

Which is generated by this function:

```
/*
 * Retina Sprites for Compass
 * by:              Gaya Kessler
 * last update:     10/25/13
 *
 * Usage:
 * 1. create two folders in your image directory (in this case 'icons' and 'icons-2x').
 * 2. adjust the foldernames defined below if you use different names.
 * 3. create sprite images for pixel ratio 1 screens and put them in the first folder.
 * 4. create sprite images for pixel ratio 2 screens and put them in the second folder, use the same filenames.
 * 5. use the sprite-image in your Sass/Scss using: '@include use-sprite(<sprite-name>)'
 */

//first we'll define the folders where the sprites are and their layouts
$icons: sprite-map("png/*.png", $layout: smart);
$icons-2x: sprite-map("png2x/*.png", $layout: smart);

//Sprite mixin, works perfectly with standard defines
@mixin use-sprite($sprite) {
    background-image: sprite-url($icons);
    background-position: sprite-position($icons, $sprite);
    background-repeat: no-repeat;
    overflow: hidden;
    display: block;
    height: image-height(sprite-file($icons, $sprite));
    width: image-width(sprite-file($icons, $sprite));

    @media (-webkit-min-device-pixel-ratio: 2), (-o-min-device-pixel-ratio: 3/2), (min--moz-device-pixel-ratio: 2), (min-device-pixel-ratio: 2), (min-resolution: 144dpi) {
        background-image: sprite-url($icons-2x);
        background-size: (image-width(sprite-path($icons-2x)) / 2) (image-height(sprite-path($icons-2x)) / 2);
        background-position: round(nth(sprite-position($icons-2x, $sprite), 1) / 2) round(nth(sprite-position($icons-2x, $sprite), 2) / 2);
        height: image-height(sprite-file($icons-2x, $sprite)) / 2;
        width: image-width(sprite-file($icons-2x, $sprite)) / 2;
    }
}
```

Config file:

```
http_path = '/'
# css_dir = 'assets/stylesheets'
# sass_dir = 'assets/sass'
images_dir = '../media/img/'
# javascripts_dir = 'assets/javascripts'
relative_assets = true
line_comments = true
# output_style = :compressed
```

**Comment by mm-tom** [comment]
I'm also seeing this, with similar code being used.

The really weird thing is, that when I roll back compass and sass I'm still getting it. Can't seem to identify what has changed at all! Infuriating.

**Comment by mm-tom** [comment]
Finally managed to fix by rolling back to alpha18 - (I hadn't rolled back properly before). Urgh. Specified:

```
gem 'compass', '< 1.0.0.alpha.19'
```

In my Gemfile which helped.

**Comment by rasky** [comment]
I've got the same bug, and cannot rollback to alpha18 because it has a different bug that was fixed in alpha19 :) I'm out of luck I guess...

**Comment by JeffreyATW** [comment]
Had to roll back to alpha18 too. Looks like alpha19 busted sprite-file.

**Comment by morewry** [comment]
I'm also getting this, and unfortunately getting https://github.com/gruntjs/grunt-contrib-compass/issues/23#issuecomment-36420740 with alpha.18 (oddly), so I'm also stuck.

**Comment by brewster1134** [comment]
alpha19 relies on the relative path from the specific sprite to your generated images path...the problem is with finding a relative path, you need to be comparing 2 absolute paths or 2 relative paths, but not 1 of each.

when the sprite is loaded, the path is expanded to the full absolute path, but the same treatment is not given to the generated images path.

a workaround for now is to ensure your `generated_images_path` is an absolute path.   you can do this by setting it explicitly, or using the default value and setting `project_path`

``` ruby
project_path: File.expand_path('.')
# or
generated_images_path: File.expand_path('./images')
```

**Comment by chriseppstein** [comment]
Set `generated_images_dir` if it's a relative location.

Duplicate of #1600

**Comment by chriseppstein** [comment]
Honestly, I cannot reproduce this unless I misconfigure my project by setting a `*_path` property to a relative url instead of an absolute path. 

http://compass-style.org/help/tutorials/configuration-reference/

If someone can provide a properly configured project that has this issue please post the details.

**Comment by brewster1134** [comment]
from what i see, the `sprite_file` function never uses `generated_images_dir`

**Comment by joeonmars** [comment]
module Compass::SassExtensions::Functions::Sprites
  ZERO = Sass::Script::Number::new(0)
  VALID_SELECTORS = %w(hover active target)
  # Provides a consistent interface for getting a variable in ruby
  # from a keyword argument hash that accounts for underscores/dash equivalence
  # and allows the caller to pass a symbol instead of a string.
  module VariableReader
    def get_var(variable_name)
      self[variable_name.to_s.gsub(/-/,"_")]
    end
  end

  #Returns a list of all sprite names
  def sprite_names(map)
    Sass::Script::List.new(map.sprite_names.map { |f| Sass::Script::String.new(f) }, ' ')
  end
  Sass::Script::Functions.declare :sprite_names, [:map]

  # Returns the system path of the sprite file
  def sprite_path(map)
    Sass::Script::String.new(map.filename)
  end
  Sass::Script::Functions.declare :sprite_path, [:map]

  # Returns the sprite file as an inline image
  #    @include "icon/*.png";
  #     #{$icon-sprite-base-class} {
  #       background-image: inline-sprite($icon-sprites);
  #      }
  def inline_sprite(map)
    verify_map(map, "sprite-url")
    map.generate
    path = map.filename
    inline_image_string(data(path), compute_mime_type(path))
  end
  Sass::Script::Functions.declare :inline_sprite, [:map]

  # Creates a Compass::SassExtensions::Sprites::SpriteMap object. A sprite map, when used in a property is the same
  # as calling sprite-url. So the following background properties are equivalent:
  #
  #     $icons: sprite-map("icons/*.png");
  #     background: sprite-url($icons) no-repeat;
  #     background: $icons no-repeat;
  #
  # The sprite map object will generate the sprite map image, if necessary,
  # the first time it is converted to a url. Simply constructing it has no side-effects.
  def sprite_map(glob, kwargs = {})
    kwargs.extend VariableReader
    Compass::SassExtensions::Sprites::SpriteMap.from_uri(glob, self, kwargs)
  end
  Sass::Script::Functions.declare :sprite_map, [:glob], :var_kwargs => true

  # Returns the image and background position for use in a single shorthand property:
  #
  #     $icons: sprite-map("icons/*.png"); // contains icons/new.png among others.
  #     background: sprite($icons, new) no-repeat;
  #
  # Becomes:
  #
  #     background: url('/images/icons.png?12345678') 0 -24px no-repeat;
  def sprite(map, sprite, offset_x = ZERO, offset_y = ZERO)
    sprite = convert_sprite_name(sprite)  
    verify_map(map)
    verify_sprite(sprite)
    url = sprite_url(map)
    position = sprite_position(map, sprite, offset_x, offset_y)
    Sass::Script::List.new([url] + position.value, :space)
  end
  Sass::Script::Functions.declare :sprite, [:map, :sprite]
  Sass::Script::Functions.declare :sprite, [:map, :sprite, :offset_x]
  Sass::Script::Functions.declare :sprite, [:map, :sprite, :offset_x, :offset_y]

  # Returns the name of a sprite map
  # The name is derived from the folder than contains the sprites.
  def sprite_map_name(map)
    verify_map(map, "sprite-map-name")
    Sass::Script::String.new(map.name)
  end
  Sass::Script::Functions.declare :sprite_name, [:sprite]

  # Returns the path to the original image file for the sprite with the given name
  def sprite_file(map, sprite)
    sprite = convert_sprite_name(sprite)
    verify_map(map, "sprite")
    verify_sprite(sprite)
    if image = map.image_for(sprite.value)
      Sass::Script::String.new(image.file)
    else
      missing_image!(map, sprite)
    end
  end
  Sass::Script::Functions.declare :sprite_file, [:map, :sprite]

  # Returns boolean if sprite has a parent
  def sprite_does_not_have_parent(map, sprite)
    sprite = convert_sprite_name(sprite)
    verify_map map
    verify_sprite sprite
    Sass::Script::Bool.new map.image_for(sprite.value).parent.nil?
  end

  Sass::Script::Functions.declare :sprite_does_not_have_parent, [:map, :sprite]

  # Returns boolean if sprite has the selector
  def sprite_has_selector(map, sprite, selector)
    sprite = convert_sprite_name(sprite)
    verify_map map
    verify_sprite sprite
    unless VALID_SELECTORS.include?(selector.value)
      raise Sass::SyntaxError, "Invalid Selctor did you mean one of: #{VALID_SELECTORS.join(', ')}"
    end
    Sass::Script::Bool.new map.send(:"has_#{selector.value}?", sprite.value)
  end

  Sass::Script::Functions.declare :sprite_has_selector, [:map, :sprite, :selector]

  # Determines if the CSS selector is valid
  def sprite_has_valid_selector(selector)
    unless selector.value =~ /\A#{Sass::SCSS::RX::IDENT}\Z/
      raise Sass::SyntaxError, "#{selector} must be a legal css identifier"
    end
    Sass::Script::Bool.new true
  end

  # Returns a url to the sprite image.
  def sprite_url(map)
    verify_map(map, "sprite-url")
    map.generate
    generated_image_url(Sass::Script::String.new("#{map.path}-s#{map.uniqueness_hash}.png"))
  end
  Sass::Script::Functions.declare :sprite_url, [:map]

  # Returns the position for the original image in the sprite.
  # This is suitable for use as a value to background-position:
  #
  #     $icons: sprite-map("icons/_.png");
  #     background-position: sprite-position($icons, new);
  #
  # Might generate something like:
  #
  #     background-position: 0 -34px;
  #
  # You can adjust the background relative to this position by passing values for
  # `$offset-x` and `$offset-y`:
  #
  #     $icons: sprite-map("icons/_.png");
  #     background-position: sprite-position($icons, new, 3px, -2px);
  #
  # Would change the above output to:
  #
  #     background-position: 3px -36px;
  def sprite_position(map, sprite = nil, offset_x = ZERO, offset_y = ZERO)
    assert_type offset_x, :Number
    assert_type offset_y, :Number
    sprite = convert_sprite_name(sprite)
    verify_map(map, "sprite-position")
    unless sprite.is_a?(Sass::Script::String) || sprite.is_a?(Sass::Script::Number)
      raise Sass::SyntaxError, %Q(The second argument to sprite-position must be a sprite name. See http://beta.compass-style.org/help/tutorials/spriting/ for more information.)
    end
    image = map.image_for(sprite.value)
    unless image
      missing_image!(map, sprite)
    end
    if offset_x.unit_str == "%"
      x = offset_x # CE: Shouldn't this be a percentage of the total width?
    else
      x = offset_x.value - image.left
      x = Sass::Script::Number.new(x, x == 0 ? [] : ["px"])
    end
    y = offset_y.value - image.top
    y = Sass::Script::Number.new(y, y == 0 ? [] : ["px"])
    Sass::Script::List.new([x, y],:space)
  end
  Sass::Script::Functions.declare :sprite_position, [:map]
  Sass::Script::Functions.declare :sprite_position, [:map, :sprite]
  Sass::Script::Functions.declare :sprite_position, [:map, :sprite, :offset_x]
  Sass::Script::Functions.declare :sprite_position, [:map, :sprite, :offset_x, :offset_y]

  def sprite_image(*args)
    raise Sass::SyntaxError, %Q(The sprite-image() function has been replaced by sprite(). See http://compass-style.org/help/tutorials/spriting/ for more information.)
  end

protected

  def reversed_color_names
    if Sass::Script::Color.const_defined?(:HTML4_COLORS_REVERSE)
      Sass::Script::Color::HTML4_COLORS_REVERSE
    else
      Sass::Script::Color::COLOR_NAMES_REVERSE
    end
  end

  def convert_sprite_name(sprite)
    case sprite
      when Sass::Script::Color
        Sass::Script::String.new(reversed_color_names[sprite.rgb])
      when Sass::Script::Bool
        Sass::Script::String.new(sprite.to_s)
      else
        sprite
    end
  end

  def verify_map(map, error = "sprite")
    unless map.is_a?(Compass::SassExtensions::Sprites::SpriteMap)
      missing_sprite!(error)
    end
  end

  def verify_sprite(sprite)
    unless sprite.is_a?(Sass::Script::String) || sprite.is_a?(Sass::Script::Number)
      raise Sass::SyntaxError, %Q(The second argument to sprite() must be a sprite name. See http://beta.compass-style.org/help/tutorials/spriting/ for more information.)
    end
  end

  def missing_image!(map, sprite)
    raise Sass::SyntaxError, "No sprite called #{sprite} found in sprite map #{map.path}/#{map.name}. Did you mean one of: #{map.sprite_names.join(", ")}"
  end

  def missing_sprite!(function_name)
    raise Sass::SyntaxError, %Q(The first argument to #{function_name}() must be a sprite map. See http://beta.compass-style.org/help/tutorials/spriting/ for more information.)
  end

end

---

Save the code above the line as 'sprites.rb' and replace the file of the same name under  /pathtoyourgems/compass-1.0.0.alpha.19/lib/compass/sass_extensions/functions

it should work. this is a persistent bug ever exist in compass older versions so maybe this could be a  quick fix to add to 1.0?
